op-e2e/sgt/sgt_test.go

```diff
@@ -0,0 +1,367 @@
+		},
+	}
+
+	for index, tCase := range tests {
+		t.Run(tCase.name, func(t *testing.T) {
+			tCase.action(t, ctx, int64(index), sgt)
+		})
+	}
+}
+
+func setUpTestAccount(t *testing.T, ctx context.Context, index int64, sgt *SgtHelper, depositSgtValue *big.Int, depositL2Value *big.Int) (*ecdsa.PrivateKey, common.Address, *big.Int) {
+	opts := &bind.CallOpts{Context: ctx}
+	rng := rand.New(rand.NewSource(index))
+	testPrivKey := testutils.InsecureRandomKey(rng)
+	testAddr := crypto.PubkeyToAddress(testPrivKey.PublicKey)
+
+	// check it's a fresh account
+	sgtBalance, err := sgt.SgtContract.BalanceOf(opts, testAddr)
+	require.NoError(t, err)
+	require.Equal(t, int64(0), sgtBalance.Int64())
+	l2Balance, err := sgt.L2Client.BalanceAt(ctx, testAddr, nil)
+	require.NoError(t, err)
+	require.Equal(t, int64(0), l2Balance.Int64())
+
+	// deposit initial sgt and native(L2) balance to the test account
+	sgt.depositSgtAndNativeFromGenesisAccountToAccount(t, ctx, testAddr, depositSgtValue, depositL2Value)
+	// ensure that sgt and native balance of testAccount are correctly initialized
+	preSgtBalance, err := sgt.SgtContract.BalanceOf(opts, testAddr)
+	require.NoError(t, err)
+	require.Equal(t, depositSgtValue.Cmp(preSgtBalance), 0)
+	preL2Balance, err := sgt.L2Client.BalanceAt(ctx, testAddr, nil)
+	require.NoError(t, err)
+	require.Equal(t, depositL2Value.Cmp(preL2Balance), 0)
+
+	return testPrivKey, testAddr, calcVaultBalance(t, ctx, sgt)
+}
+
+// balance invariant check: preTotalBalance = postTotalBalance + gasCost + txValue
+func invariantBalanceCheck(t *testing.T, ctx context.Context, sgt *SgtHelper, addr common.Address, gasCost *big.Int, txValue *big.Int, preSgtBalance *big.Int, preL2Balance *big.Int, postSgtBalance *big.Int) {
+	postL2Balance, err := sgt.L2Client.BalanceAt(ctx, addr, nil)
+	require.NoError(t, err)
+	preBalance := preSgtBalance.Add(preSgtBalance, preL2Balance)
+	postBalance := postSgtBalance.Add(postSgtBalance, gasCost)
+	postBalance = postBalance.Add(postBalance, txValue)
+	postBalance = postBalance.Add(postBalance, postL2Balance)
+	require.Equal(t, 0, preBalance.Cmp(postBalance))
+}
+
+func nativaGasPaymentWithoutSGTSuccess(t *testing.T, ctx context.Context, index int64, sgt *SgtHelper) {
+	depositSgtValue := big.NewInt(0)
+	depositL2Value := big.NewInt(10000000000000)
+	txValue := big.NewInt(0)
+	testAccount, testAddr, vaultBalanceBefore := setUpTestAccount(t, ctx, index, sgt, depositSgtValue, depositL2Value)
+
+	// make a simple tx with the testAccount: transfer txValue from testAccount to dummyAddr
+	tx, err := sgt.transferNativeToken(t, ctx, testAccount, dummyAddr, txValue)
+	require.NoError(t, err)
+	receipt, err := wait.ForReceiptOK(ctx, sgt.L2Client, tx.Hash())
+	require.NoError(t, err)
+	gasCost := calcGasFee(receipt)
+	vaultBalanceAfter := calcVaultBalance(t, ctx, sgt)
+
+	// gasCost == vaultBalanceDiff check
+	require.Equal(t, vaultBalanceAfter.Sub(vaultBalanceAfter, vaultBalanceBefore).Cmp(gasCost), 0)
```

Comment:
This will modify `vaultBalanceAfter` in place, may cause footgun sometime. Recommend: `new(big.Int).Sub...`. Same for other similar places.

Reply:
good point, done in https://github.com/ethstorage/optimism/pull/125/commits/35c926638cfc153454310ba909ca159ae103e3aa